### PR TITLE
Ask user to reauthenticate against OAuth2 server using modal dialog

### LINKED ISSUE #9559: Add "Copy to clipboard" action to "Re-open Browser" button on account screen
CleanShot 2022-04-04 at 09 27 03 | CleanShot 2022-04-04 at 09 28 30
-- | --

Could it be added there as well?

### PR DESCRIPTION
**Pull request by fmoc**
Fixes #9559.

screenshot_2022-07-04_22-28-49

### CODE REVIEW

**Comment by TheOneRing** [comment]
image

Please fix, we also discussed to display a modal dialog requesting the users attention and explaining the workflow.

File: src/gui/accountsettings.cpp [hunk 1]
```diff
@@ -797,7 +794,9 @@ void AccountSettings::slotAccountStateChanged()
                 errors << tr("The server version %1 is unsupported! Proceed at your own risk.").arg(account->capabilities().status().versionString());
             }
             showConnectionLabel(tr("Connected to %1.").arg(serverWithUser), errors);
-            ui->openBrowserButton->setVisible(false);
+            if (_askForOAuthLoginDialog != nullptr) {
+                _askForOAuthLoginDialog->accept();
```
Comment (TheOneRing) [inline src/gui/accountsettings.cpp:798]: Why accept?

File: src/gui/accountsettings.cpp [hunk 2]
```diff
@@ -812,10 +811,33 @@ void AccountSettings::slotAccountStateChanged()
...
+                qCDebug(lcAccountSettings) << "showing modal dialog asking user to log in again via OAuth2";
+
+                _askForOAuthLoginDialog = new AskForOAuthLoginDialog(_accountState->account(), this);
+
+                // make sure to clean up the memory and to null the QPointer once finished
+                _askForOAuthLoginDialog->setAttribute(Qt::WA_DeleteOnClose);
+
+                connect(
+                    cred, &HttpCredentialsGui::authorisationLinkChanged,
+                    this, [this]() {
```
Comment (TheOneRing) [inline src/gui/accountsettings.cpp:828]: Lambda not needed

**Review by TheOneRing — CHANGES_REQUESTED** [review]

**Comment by TheOneRing** [comment]
The open browser button also doesn't work

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (fmoc) [reply src/gui/accountsettings.cpp:798]: There's no particular reason. We can also just `hide()` or `close()` it if you like to. In my opinion, this makes sense to some extent, as logging in in the browser is similar to clicking a submit button in a basic auth form.

**Comment by TheOneRing** [comment]
Clicking on reopen has an effect however.
I guess you need to start the oauth flow not only create it.
Also take care of error handling, what happens if the server is not available.

**Comment by fmoc** [comment]
Cannot reproduce your reported issues. Please describe how you got there (e.g., logging out explicitly, restarting the app, etc.).

**Comment by fmoc** [comment]
I'll try the scenario "server disappears". This kind of reminds me of fixing the mess that is `authorisationLink[Async]`...

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (TheOneRing) [reply src/gui/accountsettings.cpp:798]: Accept and close will trigger the corresponding signals to be emitted.
Anyway you should add a comment that we are hiding the dialog here.
Maybe deleteLater would be correct I'm not exactly sure.

File: src/gui/accountsettings.cpp [hunk 3]
```diff
@@ -812,10 +811,31 @@ void AccountSettings::slotAccountStateChanged()
...
+                }
+
+                qCDebug(lcAccountSettings) << "showing modal dialog asking user to log in again via OAuth2";
+
+                _askForOAuthLoginDialog = new AskForOAuthLoginDialog(_accountState->account(), this);
+
+                // make sure to clean up the memory and to null the QPointer once finished
+                _askForOAuthLoginDialog->setAttribute(Qt::WA_DeleteOnClose);
+
+                connect(
```
Comment (TheOneRing) [inline src/gui/accountsettings.cpp:826]: Formated a bit weird but probably due to clang format

File: src/gui/askforoauthlogindialog.cpp [hunk 4]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace OCC {
+
+AskForOAuthLoginDialog::AskForOAuthLoginDialog(AccountPtr accountPtr, QWidget *parent)
+    : QDialog(parent)
+    , _ui(new ::Ui::AskForOAuthLoginDialog)
+{
+    _ui->setupUi(this);
+
+    _ui->label->setText(tr("The account %1 has been logged out by the server.\n\nPlease use your browser to log into %2.").arg(accountPtr->displayName(), Theme::instance()->appNameGUI()));
```
Comment (TheOneRing) [inline src/gui/askforoauthlogindialog.cpp:31]: To "login to"

File: src/gui/askforoauthlogindialog.cpp [hunk 5]
```diff
@@ -35,6 +35,15 @@ AskForOAuthLoginDialog::AskForOAuthLoginDialog(AccountPtr accountPtr, QWidget *p
 
     setFixedSize(this->sizeHint());
 
+    auto creds = qobject_cast<HttpCredentialsGui *>(accountPtr->credentials());
```
Comment (TheOneRing) [inline src/gui/askforoauthlogindialog.cpp:38]: move down to its usage

File: src/gui/askforoauthlogindialog.cpp [hunk 6]
```diff
@@ -35,6 +35,15 @@ AskForOAuthLoginDialog::AskForOAuthLoginDialog(AccountPtr accountPtr, QWidget *p
...
+
+    _ui->copyUrlToClipboardButton->setEnabled(false);
```
Comment (TheOneRing) [inline src/gui/askforoauthlogindialog.cpp:40]: Maybe disable in the ui file?

File: src/libsync/creds/oauth.cpp [hunk 7]
```diff
@@ -497,16 +497,9 @@ QUrl OAuth::authorisationLink() const
...
-}
 
-void OAuth::authorisationLinkAsync(std::function<void (const QUrl &)> callback) const
-{
-    if (_wellKnownFinished) {
-        callback(authorisationLink());
-    } else {
-        connect(this, &OAuth::authorisationLinkChanged, callback);
-    }
+    Q_ASSERT(!url.isEmpty());
```
Comment (TheOneRing) [inline src/libsync/creds/oauth.cpp:501]: How would that be possible?

Comment (TheOneRing) [reply src/libsync/creds/oauth.cpp:501]: We rather need to check for _wellKnownFinished ?

File: src/gui/askforoauthlogindialog.cpp (hunk 5 above)
Comment (fmoc) [reply src/gui/askforoauthlogindialog.cpp:38]: I totally forgot about an assertion I wanted to add there. Will fix that.

File: src/libsync/creds/oauth.cpp (hunk 7 above)
Comment (fmoc) [reply src/libsync/creds/oauth.cpp:501]: I did want to have some assertion in there at least. We can increase the checks' granularity of course.

File: src/gui/accountsettings.cpp (hunk 1 above)
Comment (TheOneRing) [reply src/gui/accountsettings.cpp:798]: if hide deletes the dialog :D

File: src/gui/askforoauthlogindialog.cpp (hunk 4 above)
Comment (fmoc) [reply src/gui/askforoauthlogindialog.cpp:31]: I used `log into` everywhere else. I assume we have to adapt those occurrences as well then.

Comment (TheOneRing) [reply src/gui/askforoauthlogindialog.cpp:31]: Hmm yes  but we should discuss it in chat maybe we can find a native speaker

**Pull request merged**